src/vm/types/serialization.rs

```diff
@@ -1,369 +1,467 @@
-                let type_name = if response.committed {
-                    TYPE_RESP_OK
+                if response.committed {
+                    TypePrefix::ResponseOk
                 } else {
-                    TYPE_RESP_ERR
-                };
-                let value = response.data.serialize();
-                json_recursive_object(type_name, &value)
-            },
-            Optional(OptionalData{ data: None }) => {
-                format!(r#"{{ "type": "{}" }}"#, TYPE_OPT_NONE)
-            },
-            Optional(OptionalData{ data: Some(value) }) => {
-                json_recursive_object(TYPE_OPT_SOME, &value.serialize()) 
-           },
-            List(data) => {
-                let entries: Vec<String> = data.data
-                    .iter().map(|x| x.serialize())
-                    .collect();
-                let entries_str = entries.join(", ");
-                format!(
-                    r#"{{ "type": "{}", "entries": [ {} ] }}"#,
-                    TYPE_LIST, entries_str)
+                    TypePrefix::ResponseErr
+                }
             },
-            Tuple(data) => {
-                let entries: Vec<String> = data.data_map
-                    .iter().map(|(key, value)|
-                                format!(r#""{}": {}"#, &**key, value.serialize()))
-                    .collect();
-                let entries_str = entries.join(", ");
-                format!(
-                    r#"{{ "type": "{}", "entries": {{ {} }} }}"#,
-                    TYPE_TUPLE, entries_str)
-            }
+            Optional(OptionalData{ data: None }) => TypePrefix::OptionalNone,
+            Optional(OptionalData{ data: Some(value) }) => TypePrefix::OptionalSome,
+            List(_) => TypePrefix::List,
+            Tuple(_) => TypePrefix::Tuple
         }
     }
 }
 
+/// Not a public trait,
+///   this is just used to simplify serializing some types that
+///   are repeatedly serialized or deserialized.
+trait ClarityValueSerializable<T: std::marker::Sized> {
+    fn serialize_write<W: Write>(&self, w: &mut W) -> std::io::Result<()>;
+    fn deserialize_read<R: Read>(r: &mut R) -> Result<T, SerializationError>;
+}
+
+impl ClarityValueSerializable<StandardPrincipalData> for StandardPrincipalData {
+    fn serialize_write<W: Write>(&self, w: &mut W) -> std::io::Result<()> {
+        w.write_all(&[self.0])?;
+        w.write_all(&self.1)
+    }
+
+    fn deserialize_read<R: Read>(r: &mut R) -> Result<Self, SerializationError> {
+        let mut version = [0; 1];
+        let mut data = [0; 20];
+        r.read_exact(&mut version)?;
+        r.read_exact(&mut data)?;
+        Ok(StandardPrincipalData(version[0], data))
+    }
+}
+
+macro_rules! serialize_guarded_string {
+    ($Name:ident) => {
+
+impl ClarityValueSerializable<$Name> for $Name {
+    fn serialize_write<W: Write>(&self, w: &mut W) -> std::io::Result<()> {
+        w.write_all(&u32::try_from(self.as_str().len())
```

Comment:
I thought a guarded string type (`ClarityName`, `ContractName`) could never be more than 128 bytes?  I need clarity (no pun intended) on this either way, because the transaction serialization code assumes this.

Reply:
Yes, it is -- will change this to a u8.

Reply:
Changed in 388ee32